smack-experimental/src/main/java/org/jivesoftware/smackx/httpfileupload/HttpFileUploadManager.java

```diff
@@ -0,0 +1,213 @@
+        if (dataForm == null) {
+            return uploadService;
+        }
+
+        FormField field = dataForm.getField("max-file-size");
+        if (field == null) {
+            return uploadService;
+        }
+
+        List<String> values = field.getValues();
+        if (!values.isEmpty()) {
+            // TODO: catch NumberFormatException?
+            maxFileSize = Long.valueOf(values.get(0));
+        }
+
+        return uploadService;
+    }
+
+    /**
+     * Get upload service if available.
+     *
+     * Note that you must call {@link HttpFileUploadManager#discoverUploadService} first.
+     *
+     * @return upload service JID or null if not available
+     */
+    public DomainBareJid getUploadService() {
+        return uploadService;
+    }
+
+    /**
+     * Get max file size allowed by upload service.
+     *
+     * Note that you must call {@link HttpFileUploadManager#discoverUploadService} first.
+     *
+     * @return max file size in bytes or null if not available
+     */
+    public Long getMaxFileSize() {
+        return maxFileSize;
+    }
+
+    /**
+     * Request a new upload slot.
+     *
+     * Note that this is a synchronous call -- Smack must wait for the server response.
+     *
+     * @param filename name of file to be uploaded
+     * @param fileSize file size in bytes -- must be less or equal to {@link HttpFileUploadManager#getMaxFileSize()} (if available)
+     * @return file upload Slot in case of success
+
+     * @throws IllegalArgumentException if fileSize is greater than {@link HttpFileUploadManager#getMaxFileSize()}
+     * @throws InterruptedException
+     * @throws XMPPException.XMPPErrorException
+     * @throws SmackException.NotConnectedException
+     * @throws SmackException.NoResponseException
+     */
+    public Slot requestSlot(String filename, long fileSize) throws InterruptedException,
+            XMPPException.XMPPErrorException, SmackException.NotConnectedException, SmackException.NoResponseException {
+        return requestSlot(filename, fileSize, null);
+    }
+
+    /**
+     * Request a new upload slot.
+     *
+     * Note that this is a synchronous call -- Smack must wait for the server response.
+     *
+     * @param filename name of file to be uploaded
+     * @param fileSize file size in bytes -- must be less or equal to {@link HttpFileUploadManager#getMaxFileSize()} (if available)
+     * @param contentType file content-type or null
+     * @return file upload Slot in case of success
+
+     * @throws IllegalArgumentException if fileSize is greater than {@link HttpFileUploadManager#getMaxFileSize()}
+     * @throws SmackException.NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPException.XMPPErrorException
+     * @throws SmackException.NoResponseException
+     */
+    public Slot requestSlot(String filename, long fileSize, String contentType) throws SmackException.NotConnectedException,
```

Comment:
I'd also like to have a method `requestSlot(String, long, String, DomainBareJid)` where the user can specify the upload service to be used.